[PropertyInfo] PhpStan FQSEN error

### Symfony version(s) affected

5.4, 6.*

### Description

PhpStanExtractor incorrect detect class with different namespace. PhpDocExtractor work correct.

#40457

### How to reproduce

```php
<?php

namespace {
    require_once __DIR__ . '/vendor/autoload.php';

    use Symfony\Component\PropertyInfo\Extractor\PhpDocExtractor;
    use Symfony\Component\PropertyInfo\Extractor\PhpStanExtractor;

    $phpDocExtractor = new PhpDocExtractor();
    $phpStanExtractor = new PhpStanExtractor();


    dump(
        $phpDocTypes = $phpDocExtractor->getTypes(B\Dummy::class, 'property'),
        $phpStanTypes = $phpStanExtractor->getTypes(B\Dummy::class, 'property')
    );

    assert($phpStanTypes[0]->getClassName() === 'A\Property');
    assert($phpDocTypes[0]->getClassName() === $phpStanTypes[0]->getClassName());
}


namespace A {
    class Property {

    }

    class Dummy {
        /**
         * @var Property
         */
        public $property;
    }
}

namespace B {
    class Dummy extends \A\Dummy {

    }
}

```

### Possible Solution

_No response_

### Additional Context

_No response_

Yes, i ran this code on version property-info v5.4.3
Output:
```
array:1 [
  0 => Symfony\Component\PropertyInfo\Type {#20
    -builtinType: "object"
    -nullable: false
    -class: "A\Property"
    -collection: false
    -collectionKeyType: []
    -collectionValueType: []
  }
]
array:1 [
  0 => Symfony\Component\PropertyInfo\Type {#32
    -builtinType: "object"
    -nullable: false
    -class: "B\Property"
    -collection: false
    -collectionKeyType: []
    -collectionValueType: []
  }
]

Fatal error: Uncaught AssertionError: assert($phpStanTypes[0]->getClassName() === 'A\\Property') in /Users/andrejisypov/Projects/ucheba/ucheba-backend/src/test.php:17
Stack trace:
#0 /Users/andrejisypov/Projects/ucheba/ucheba-backend/src/test.php(17): assert(false, 'assert($phpStan...')
#1 {main}
  thrown in /Users/andrejisypov/Projects/ucheba/ucheba-backend/src/test.php on line 17
```

--- a/src/Symfony/Component/PropertyInfo/Extractor/PhpStanExtractor.php
+++ b/src/Symfony/Component/PropertyInfo/Extractor/PhpStanExtractor.php
@@ -45,7 +45,7 @@ final class PhpStanExtractor implements PropertyTypeExtractorInterface, Construc
     /** @var NameScopeFactory */
     private $nameScopeFactory;
 
-    /** @var array<string, array{PhpDocNode|null, int|null, string|null}> */
+    /** @var array<string, array{PhpDocNode|null, int|null, string|null, string|null}> */
     private $docBlocks = [];
     private $phpStanTypeHelper;
     private $mutatorPrefixes;
@@ -72,8 +72,8 @@ final class PhpStanExtractor implements PropertyTypeExtractorInterface, Construc
     public function getTypes(string $class, string $property, array $context = []): ?array
     {
         /** @var PhpDocNode|null $docNode */
-        [$docNode, $source, $prefix] = $this->getDocBlock($class, $property);
-        $nameScope = $this->nameScopeFactory->create($class);
+        [$docNode, $source, $prefix, $declaringClass] = $this->getDocBlock($class, $property);
+        $nameScope = $this->nameScopeFactory->create($class, $declaringClass);
         if (null === $docNode) {
             return null;
         }
@@ -184,7 +184,7 @@ final class PhpStanExtractor implements PropertyTypeExtractorInterface, Construc
     }
 
     /**
-     * @return array{PhpDocNode|null, int|null, string|null}
+     * @return array{PhpDocNode|null, int|null, string|null, string|null}
      */
     private function getDocBlock(string $class, string $property): array
     {
@@ -196,20 +196,23 @@ final class PhpStanExtractor implements PropertyTypeExtractorInterface, Construc
 
         $ucFirstProperty = ucfirst($property);
 
-        if ($docBlock = $this->getDocBlockFromProperty($class, $property)) {
-            $data = [$docBlock, self::PROPERTY, null];
-        } elseif ([$docBlock] = $this->getDocBlockFromMethod($class, $ucFirstProperty, self::ACCESSOR)) {
-            $data = [$docBlock, self::ACCESSOR, null];
-        } elseif ([$docBlock, $prefix] = $this->getDocBlockFromMethod($class, $ucFirstProperty, self::MUTATOR)) {
-            $data = [$docBlock, self::MUTATOR, $prefix];
+        if ([$docBlock, $declaringClass] = $this->getDocBlockFromProperty($class, $property)) {
+            $data = [$docBlock, self::PROPERTY, null, $declaringClass];
+        } elseif ([$docBlock, $_, $declaringClass] = $this->getDocBlockFromMethod($class, $ucFirstProperty, self::ACCESSOR)) {
+            $data = [$docBlock, self::ACCESSOR, null, $declaringClass];
+        } elseif ([$docBlock, $prefix, $declaringClass] = $this->getDocBlockFromMethod($class, $ucFirstProperty, self::MUTATOR)) {
+            $data = [$docBlock, self::MUTATOR, $prefix, $declaringClass];
         } else {
-            $data = [null, null, null];
+            $data = [null, null, null, null];
         }
 
         return $this->docBlocks[$propertyHash] = $data;
     }
 
-    private function getDocBlockFromProperty(string $class, string $property): ?PhpDocNode
+    /**
+     * @return array{PhpDocNode, string}|null
+     */
+    private function getDocBlockFromProperty(string $class, string $property): ?array
     {
         // Use a ReflectionProperty instead of $class to get the parent class if applicable
         try {
@@ -226,11 +229,11 @@ final class PhpStanExtractor implements PropertyTypeExtractorInterface, Construc
         $phpDocNode = $this->phpDocParser->parse($tokens);
         $tokens->consumeTokenType(Lexer::TOKEN_END);
 
-        return $phpDocNode;
+        return [$phpDocNode, $reflectionProperty->class];
     }
 
     /**
-     * @return array{PhpDocNode, string}|null
+     * @return array{PhpDocNode, string, string}|null
      */
     private function getDocBlockFromMethod(string $class, string $ucFirstProperty, int $type): ?array
     {
@@ -269,6 +272,6 @@ final class PhpStanExtractor implements PropertyTypeExtractorInterface, Construc
         $phpDocNode = $this->phpDocParser->parse($tokens);
         $tokens->consumeTokenType(Lexer::TOKEN_END);
 
-        return [$phpDocNode, $prefix];
+        return [$phpDocNode, $prefix, $reflectionMethod->class];
     }
 }

--- a/src/Symfony/Component/PropertyInfo/PhpStan/NameScope.php
+++ b/src/Symfony/Component/PropertyInfo/PhpStan/NameScope.php
@@ -22,14 +22,14 @@ namespace Symfony\Component\PropertyInfo\PhpStan;
  */
 final class NameScope
 {
-    private $className;
+    private $calledClassName;
     private $namespace;
     /** @var array<string, string> alias(string) => fullName(string) */
     private $uses;
 
-    public function __construct(string $className, string $namespace, array $uses = [])
+    public function __construct(string $calledClassName, string $namespace, array $uses = [])
     {
-        $this->className = $className;
+        $this->calledClassName = $calledClassName;
         $this->namespace = $namespace;
         $this->uses = $uses;
     }
@@ -60,6 +60,6 @@ final class NameScope
 
     public function resolveRootClass(): string
     {
-        return $this->resolveStringName($this->className);
+        return $this->resolveStringName($this->calledClassName);
     }
 }

--- a/src/Symfony/Component/PropertyInfo/PhpStan/NameScopeFactory.php
+++ b/src/Symfony/Component/PropertyInfo/PhpStan/NameScopeFactory.php
@@ -20,16 +20,18 @@ use phpDocumentor\Reflection\Types\ContextFactory;
  */
 final class NameScopeFactory
 {
-    public function create(string $fullClassName): NameScope
+    public function create(string $calledClassName, string $declaringClassName = null): NameScope
     {
-        $reflection = new \ReflectionClass($fullClassName);
-        $path = explode('\\', $fullClassName);
-        $className = array_pop($path);
-        [$namespace, $uses] = $this->extractFromFullClassName($reflection);
+        $declaringClassName = $declaringClassName ?? $calledClassName;
 
-        $uses = array_merge($uses, $this->collectUses($reflection));
+        $path = explode('\\', $calledClassName);
+        $calledClassName = array_pop($path);
 
-        return new NameScope($className, $namespace, $uses);
+        $declaringReflection = new \ReflectionClass($declaringClassName);
+        [$declaringNamespace, $declaringUses] = $this->extractFromFullClassName($declaringReflection);
+        $declaringUses = array_merge($declaringUses, $this->collectUses($declaringReflection));
+
+        return new NameScope($calledClassName, $declaringNamespace, $declaringUses);
     }
 
     private function collectUses(\ReflectionClass $reflection): array

--- a/src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpStanExtractorTest.php
+++ b/src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpStanExtractorTest.php
@@ -12,6 +12,7 @@
 namespace Symfony\Component\PropertyInfo\Tests\Extractor;
 
 use PHPUnit\Framework\TestCase;
+use Symfony\Component\PropertyInfo\Extractor\PhpDocExtractor;
 use Symfony\Component\PropertyInfo\Extractor\PhpStanExtractor;
 use Symfony\Component\PropertyInfo\Tests\Fixtures\DefaultValue;
 use Symfony\Component\PropertyInfo\Tests\Fixtures\Dummy;
@@ -20,6 +21,8 @@ use Symfony\Component\PropertyInfo\Tests\Fixtures\RootDummy\RootDummyItem;
 use Symfony\Component\PropertyInfo\Tests\Fixtures\TraitUsage\DummyUsedInTrait;
 use Symfony\Component\PropertyInfo\Tests\Fixtures\TraitUsage\DummyUsingTrait;
 use Symfony\Component\PropertyInfo\Type;
+
+require_once __DIR__.'/../Fixtures/Extractor/DummyNamespace.php';
 
 /**
  * @author Baptiste Leduc <baptiste.leduc@gmail.com>
@@ -31,9 +34,15 @@ class PhpStanExtractorTest extends TestCase
      */
     private $extractor;
 
+    /**
+     * @var PhpDocExtractor
+     */
+    private $phpDocExtractor;
+
     protected function setUp(): void
     {
         $this->extractor = new PhpStanExtractor();
+        $this->phpDocExtractor = new PhpDocExtractor();
     }
 
     /**
@@ -382,6 +391,15 @@ class PhpStanExtractorTest extends TestCase
             [new Type(Type::BUILTIN_TYPE_OBJECT, false, 'Symfony\Component\PropertyInfo\Tests\Fixtures\Dummy')],
             $this->extractor->getTypes('Symfony\Component\PropertyInfo\Tests\Fixtures\DummyNamespace', 'dummy')
         );
+    }
+
+    public function testDummyNamespaceWithProperty()
+    {
+        $phpStanTypes = $this->extractor->getTypes(\B\Dummy::class, 'property');
+        $phpDocTypes = $this->phpDocExtractor->getTypes(\B\Dummy::class, 'property');
+
+        $this->assertEquals('A\Property', $phpStanTypes[0]->getClassName());
+        $this->assertEquals($phpDocTypes[0]->getClassName(), $phpStanTypes[0]->getClassName());
     }
 }
 

--- a/src/Symfony/Component/PropertyInfo/Tests/Fixtures/Extractor/DummyNamespace.php
+++ b/src/Symfony/Component/PropertyInfo/Tests/Fixtures/Extractor/DummyNamespace.php
@@ -1,0 +1,20 @@
+<?php
+
+namespace A {
+    class Property {
+
+    }
+
+    class Dummy {
+        /**
+         * @var Property
+         */
+        public $property;
+    }
+}
+
+namespace B {
+    class Dummy extends \A\Dummy {
+
+    }
+}

